### PR TITLE
Building Windows image

### PR DESCRIPTION
**Pull request by hairyhenderson**
This adds a build for a windows flavour of the Caddy Docker image!

Signed-off-by: Dave Henderson <dhenderson@gmail.com>

### CODE REVIEW

File: windows/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+LABEL org.opencontainers.image.documentation=https://caddyserver.com/docs
+LABEL org.opencontainers.image.vendor="Light Code Labs"
+LABEL org.opencontainers.image.licenses=Apache-2.0
+LABEL org.opencontainers.image.source="https://github.com/caddyserver/caddy-docker"
+
+EXPOSE 80
+EXPOSE 443
+EXPOSE 2019
+
+CMD ["caddy", "run", "--config", "/etc/caddy/Caddyfile", "--adapter", "caddyfile"]
```
Comment (mholt) [inline windows/Dockerfile:47]: FWIW, `--adapter caddyfile` isn't necessary when the `--config` filename starts with `Caddyfile`.

Comment (francislavoie) [reply windows/Dockerfile:47]: Doesn't hurt to be explicit though IMO 🤷‍♂️

Comment (mholt) [reply windows/Dockerfile:47]: It just means that if the user doesn't use a Caddyfile and wants to use JSON, they have to remember to remove the `--adapter` flag, even though they're using Caddy's native format. I don't care too much either way, just wanted to point this out. I don't want people thinking they're required to use that flag for the Caddyfile all the time.

Comment (francislavoie) [reply windows/Dockerfile:47]: When you run with docker, `CMD` is just the default command. If you specify a different one, it just overrides the existing one entirely. The user wouldn't need to remove anything if they run with JSON.

e.g. `docker run caddy caddy run --config /etc/caddy/config.json`

Comment (hairyhenderson) [reply windows/Dockerfile]: This is likely to change too - the image is very much not working in Windows right now 😉

File: Dockerfile.windows.tmpl [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+EXPOSE 80
+EXPOSE 443
+EXPOSE 2019
+
+# CMD ["caddy", "run", "--config", "/etc/caddy/Caddyfile", "--adapter", "caddyfile"]
+
+FROM caddy AS test
+
+RUN /caddy.exe --version
```
Comment (francislavoie) [inline Dockerfile.windows.tmpl:51]:
```suggestion
RUN /caddy.exe version
```

**Comment by hairyhenderson** [comment]
Ok, this works for a basic site with the default config now. I've fired up a Win2019 VM on Azure and tested it out.

Would be useful to also test out the automatic TLS stuff, just to be sure. Then I think it's probably ready to merge.

File: windows/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        -OutFile "/caddy.zip"; \
+    Expand-Archive -Path "/caddy.zip" -DestinationPath "/" -Force; \
+    Remove-Item "/caddy.zip" -Force
+
+# See https://caddyserver.com/docs/conventions#file-locations for details
+# ENV XDG_CONFIG_HOME=/config
+# ENV XDG_DATA_HOME=/data
+
+# VOLUME /config
+# VOLUME /data
```
Comment (francislavoie) [inline windows/Dockerfile:31]: What do the volumes look like on Windows then?

Comment (hairyhenderson) [reply windows/Dockerfile]: er... sorry, I committed but never pushed 😂 - this is out of date!

File: .github/workflows/docker.yml [hunk 4]
```diff
@@ -30,3 +30,26 @@ jobs:
...
+    # env:
+    #   DOCKER_BUILDKIT: '1'
+    steps:
+      - uses: actions/checkout@master
+      - name: non-master build test
+        run: |
+          docker build -f windows/Dockerfile windows
+        if: github.repository != 'caddyserver/caddy-docker' || github.ref != 'refs/heads/master'
+      # - name: build
+      #   run: bashbrew build caddy
```
Comment (hairyhenderson) [inline .github/workflows/docker.yml:44]: Not sure if we can get this to run (easily) in Windows... I think it's enough to let the official build servers publish the windows flavour.

**Comment by hairyhenderson** [comment]
Another outstanding question... I'd like to verify the zip, but not sure how in Windows... Perhaps there's some powershell magic for this? I'll have another look tomorrow evening...

File: windows/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+LABEL org.opencontainers.image.source="https://github.com/caddyserver/caddy-docker"
+
+EXPOSE 80
+EXPOSE 443
+EXPOSE 2019
+
+# Make sure it runs and reports its version
+RUN ["caddy", "version"]
+
+CMD ["caddy", "run", "--config", "/caddy/Caddyfile", "--adapter", "caddyfile"]
```
Comment (francislavoie) [inline windows/Dockerfile:51]: Looks like the default Caddyfile is in a different path than on the linux variants. Kinda less fun to document that. Can we unify that?

Comment (hairyhenderson) [reply windows/Dockerfile:51]: Yeah, I waffled on that for a bit... We can definitely put it in `/etc/caddy/`. It translates to `C:\etc\caddy\` in Windows, but `/` works fine 😉

**Comment by hairyhenderson** [comment]
> ...but not sure how in Windows... Perhaps there's some powershell magic for this?

Indeed! `Get-FileHash` is the key 😬

**Comment by hairyhenderson** [comment]
I've tested this out with a custom `Caddyfile`:

```
caddywindows.hairyhenderson.ca

root * /usr/share/caddy
file_server
```

```
PS C:\Users\hairyhenderson\caddy> docker run -it --rm -p 80:80 -p 443:443 -v caddy_data:c:\data -v caddy_config:c:\config -v c:/Users/hairyhenderson/caddy:c:/etc/caddy caddy
```

And it worked just fine:

```
2020/04/17 23:45:16.826 INFO   using provided configuration    {"config_file": "/etc/caddy/Caddyfile", "config_adapter": "caddyfile"}
2020/04/17 23:45:16.835 INFO   admin   admin endpoint started  {"address": "tcp/localhost:2019", "enforce_origin": false, "origins": ["localhost:2019", "[::1]:2019", "127.0.0.1:2019"]}
2020/04/17 23:45:16.836 INFO   http    server is listening only on the HTTPS port but has no TLS connection policies; adding one to enable TLS {"server_name": "srv0", "https_port": 443}
2020/04/17 23:45:16.839 INFO   http    enabling automatic HTTP->HTTPS redirects        {"server_name": "srv0"}
2020/04/17 23:45:16.840 INFO   tls     cleaned up storage units
2020/04/17 23:45:16.841 INFO   http    enabling automatic TLS certificate management   {"domains": ["caddywindows.hairyhenderson.ca"]}
2020/04/17 23:45:16.942 INFO   autosaved config        {"file": "c:\\config\\caddy\\autosave.json"}
2020/04/17 23:45:16.943 INFO   serving initial configuration
2020/04/17 23:45:16 [INFO][cache:0xc0006f3a90] Started certificate maintenance routine
2020/04/17 23:45:17 [INFO][caddywindows.hairyhenderson.ca] Obtain certificate; acquiring lock...
2020/04/17 23:45:17 [INFO][caddywindows.hairyhenderson.ca] Obtain: Lock acquired; proceeding...
2020/04/17 23:45:18 [INFO][caddywindows.hairyhenderson.ca] Waiting on rate limiter...
2020/04/17 23:45:18 [INFO][caddywindows.hairyhenderson.ca] Done waiting
2020/04/17 23:45:18 [INFO] [caddywindows.hairyhenderson.ca] acme: Obtaining bundled SAN certificate given a CSR
2020/04/17 23:45:18 [INFO] [caddywindows.hairyhenderson.ca] AuthURL: https://acme-v02.api.letsencrypt.org/acme/authz-v3/xxxxxxxxxx
2020/04/17 23:45:18 [INFO] [caddywindows.hairyhenderson.ca] acme: Could not find solver for: tls-alpn-01
2020/04/17 23:45:18 [INFO] [caddywindows.hairyhenderson.ca] acme: use http-01 solver
2020/04/17 23:45:18 [INFO] [caddywindows.hairyhenderson.ca] acme: Trying to solve HTTP-01
2020/04/17 23:45:18 [INFO][caddywindows.hairyhenderson.ca] Served key authentication (HTTP challenge)
2020/04/17 23:45:19 [INFO][caddywindows.hairyhenderson.ca] Served key authentication (HTTP challenge)
2020/04/17 23:45:19 [INFO][caddywindows.hairyhenderson.ca] Served key authentication (HTTP challenge)
2020/04/17 23:45:19 [INFO][caddywindows.hairyhenderson.ca] Served key authentication (HTTP challenge)
2020/04/17 23:45:21 [INFO] [caddywindows.hairyhenderson.ca] The server validated our request
2020/04/17 23:45:21 [INFO] [caddywindows.hairyhenderson.ca] acme: Validations succeeded; requesting certificates
2020/04/17 23:45:22 [INFO] [caddywindows.hairyhenderson.ca] Server responded with a certificate.
2020/04/17 23:45:22 [INFO][caddywindows.hairyhenderson.ca] Certificate obtained successfully
2020/04/17 23:45:22 [INFO][caddywindows.hairyhenderson.ca] Obtain: Releasing lock
```

**Comment by hairyhenderson** [comment]
(note: the rebase/squash was so that I could remove the temporary `library/caddy` stackfile - that'll go in once this merges to `master`)

**Comment by hairyhenderson** [comment]
The build failure for `docker-windows-build` is expected - it'll pass once this merges and the new stackfile is committed! 😉

**Pull request merged**